src/event/metric.rs

```diff
@@ -57,4 +58,85 @@ impl Metric {
             Metric::Set { tags, .. } => tags,
         }
     }
+
+    pub fn merge(&mut self, other: &Metric) {
+        match (self, other) {
+            (
+                Metric::Counter {
+                    ref mut name,
+                    ref mut val,
+                    ref mut timestamp,
+                    ref mut tags,
+                },
+                Metric::Counter {
+                    name: new_name,
+                    val: new_val,
+                    timestamp: new_timestamp,
+                    tags: new_tags,
+                },
+            ) => {
+                *val += *new_val;
+                *name = new_name.clone();
+                *timestamp = *new_timestamp;
+                *tags = new_tags.clone();
+            }
+            (
+                Metric::Gauge {
+                    ref mut name,
+                    ref mut val,
+                    direction: None,
+                    ref mut timestamp,
+                    ref mut tags,
+                },
+                Metric::Gauge {
+                    name: new_name,
+                    val: new_val,
+                    timestamp: new_timestamp,
+                    direction: new_direction,
+                    tags: new_tags,
+                },
+            ) => {
+                if new_direction.is_none() {
+                    *val = *new_val;
+                } else {
+                    let delta = match new_direction {
+                        None => 0.0,
+                        Some(Direction::Plus) => *val,
+                        Some(Direction::Minus) => -*val,
+                    };
+                    *val += delta;
+                };
+                *name = new_name.clone();
+                *timestamp = *new_timestamp;
+                *tags = new_tags.clone();
+            }
+            (
+                Metric::Histogram {
+                    ref mut name,
+                    ref mut val,
+                    ref mut sample_rate,
+                    ref mut timestamp,
+                    ref mut tags,
+                },
+                Metric::Histogram {
+                    name: new_name,
+                    val: new_val,
+                    sample_rate: new_sample_rate,
+                    timestamp: new_timestamp,
+                    tags: new_tags,
+                },
+            ) => {
+                if val == new_val {
```

Comment:
I agree that we might expect a lot of variation in _timings_ data, especially if its granular (ms).

But there is also another big domain of _size_ data. Like `Bytes Out`, `Content Size` kind of things, which are going to have a lot of repetition actually, especially in heavy-read workloads.

Also, if the system being logged has distinction between _processed_ and _billed_ (like [BigQuery](https://cloud.google.com/bigquery/docs/reference/auditlogs/rest/Shared.Types/BigQueryAuditMetadata) for example), then with caching enabled it will produce a lot of same values (zeros) in fields like `totalBilledBytes`.

This _processed_ VS _billed_ dichotomy can also affect timings data. Let's take [AWS Lambda](https://aws.amazon.com/blogs/mt/understanding-aws-lambda-behavior-using-amazon-cloudwatch-logs-insights/) logs for example.

```
REPORT RequestId: 8b4a32a4-7dc0-4342-a998-48b051d683e0 Duration: 182.29 ms Billed Duration: 200 ms Memory Size: 128 MB Max Memory Used: 35 MB
```

The `Billed Duration` thing here will be highly repetitive.

Reply:
> If we want to support histograms, we'll need something more general than this.

The nice feature of this `sample_rate` hack is that it is a lossless transformation. I agree that having client-size (lossy) histograms aggregation is important, I just think we need a separate issue for it.

Reply:
My original idea was to expand the `Histogram` variant to be able to hold more than one data point, so merging would be as simple as merging the sets of points. I'm curious how dramatic of a change that would be across the rest of the code base.

Reply:
But looking at the cloudwatch docs, it doesn't seem like that's actually a model that they support in the API.